KEP for co-scheduling plugin based on scheduler framework (#1463)
kubernetes/enhancements: keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

#### Permit
In `Permit` phase, we put the pod that doesn't meet min-available into the WaitingMap and reserve resources until min-available are met or timeout is triggered.
1. Get the number of Running pods that belong to the same PodGroup
2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: the timeout is dynamic value depends on the size of the `PodGroup`.)

#### UnReserve
After a pod which belongs to a PodGroup times out in the permit phase.  UnReserve ```Rejects``` the pods that belong to the same PodGroup to avoid long-term invalid reservation of resources.

## Alternatives considered
1. Using `PodGroup` as a scheduling unit. This requires major refactoring, which only supports Pods as scheduling unit today.

## Graduation Criteria

## Testing Plan
1.  Add detailed unit and integration tests for workloads.
2.  Add basic e2e tests, to ensure all components are working together.

>  

Reviewer: (as someone who mainly contributes to SIG Docs) I like to see at least a sketch of documentation requirements for feature graduation.

Would that be OK?

```go
func  Less(podA *PodInfo, podB *PodInfo) bool
```

Firstly, we will inherit the default in-tree PrioritySort plugin so as to honor .spec.priority to ensure high-priority Pods are always sorted ahead of low-priority ones.

Secondly, if two Pods hold the same priority, the sorting precedence is described as below:

- If they are both regularPods (without particular PodGroup label), compare their `InitialAttemptTimestamp` field: the Pod with earlier `InitialAttemptTimestamp` is positioned ahead of the other.
  
- If one is regularPod and the other is pgPod, compare regularPod's `InitialAttemptTimestamp` with pgPod's `InitialTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
  
- If they are both pgPods:
  - Compare their `InitialAttemptTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
  - If their `InitialAttemptTimestamp` is identical, order by their UID of PodGroup: a Pod with lexicographically greater UID is scheduled ahead of the other Pod. (The purpose is to tease different PodGroups with the same `InitialAttemptTimestamp` apart, while also keeping Pods belonging to the same PodGroup back-to-back)

#### Pre-Filter

> When a `pgGroup` Pod is being scheduled for the first time, we have 2 option to deal with it: either start the full scheduling cycle no matter its grouping Pods are present inside schedule queue, or fail quick its scheduling cycle as its grouping Pods number doesn't meet `minAvailable`. The former case is more efficient, but may cause partial Pods holding system resources until a timeout. The latter case may result in extra scheduling cycles (even if we're going to fail them fast), but will reduce the overall scheduling time for the whole group - as we're waiting them to be all present in the queue first and then start the full scheduling cycle for each).

Reviewer 2: s/pgGroup Pod/pgPod ?

Author: done

Here we're adopting the latter approach, `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.

#### Permit
In `Permit` phase, we put the pod that doesn't meet `minAvailable` into the WaitingMap and reserve resources until `minAvailable` are met or timeout is triggered.
1. Get the number of Running pods that belong to the same PodGroup
2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup

> 3. If Running + WaitingPods + 1 >= `minAvailable`(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: the timeout is dynamic value depends on the size of the `PodGroup`.)

Reviewer 2: I think of a generation algorithm about the dynamic timeout. We know that a group of pods is exactly the same. When the cluster node is unchanged, we can assume that the scheduling time of each pod is the same, then we can get a formula 
 `(minAvailable-waitingPods -1) * schedulingTime`, where `schedulingTime` is the time that pod joins the scheduling queue to run to the `permit` plug-in. For example, if the `minAvailable` of a job is `5` and the time to schedule a pod is `10ms`, then the waiting time for the first pod is `(5-0-1) * 10ms`, the waiting time for the second pod is `(5-1-1) * 10ms`, and so on for the rest of the pods. Considering network, load, etc., it is possible to add a random time to the timeout. **WDYT**?

> 3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: the timeout is dynamic value depends on the size of the `PodGroup`.)

Reviewer 2: s/min-available/MinAvailable, the former is easily misunderstood as min minus available

Author: done

authors:
  - "@denkensk"
owning-sig: sig-scheduling
reviewers:
  - "@Huang-Wei"
  - "@ahg-g"
  - "@alculquicondor"
  - "k82cn"
  - "@resouer"
  - "@hex108"
  - "@everpeace"
approvers:
  - "@Huang-Wei"
creation-date: 2020-01-16
last-updated: 2020-01-16
status: provisional
---

# Coscheduling plugin based on scheduler framework

## Table of Contents

<!-- toc -->
- [Coscheduling plugin based on scheduler framework](#coscheduling-plugin-based-on-scheduler-framework)
  - [Table of Contents](#table-of-contents)
  - [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
  - [Use Cases](#use-cases)
  - [Terms](#terms)
  - [Proposal](#proposal)
  - [Design Details](#design-details)
    - [PodGroup](#podgroup)
    - [Coscheduling](#coscheduling)
    - [Extension points](#extension-points)
      - [QueueSort](#queuesort)
      - [Pre-Filter](#pre-filter)
      - [Permit](#permit)
      - [UnReserve](#unreserve)
  - [Alternatives considered](#alternatives-considered)
  - [Graduation Criteria](#graduation-criteria)
  - [Testing Plan](#testing-plan)
  - [Implementation History](#implementation-history)
  - [References](#references)
<!-- /toc -->

## Motivation

> Kubernetes has become a popular solution for orchestrating containerized workloads. Due to limitation of Kubernetes scheduler, some offline workloads (ML/DL) are managed in a different system. To improve cluster utilization and operation efficiency, we'd like to treat Kubneretes as a unified management platform. But ML jobs are All-or-Nothing: they require all tasks of a job to be scheduled at the same time. If the job only start part of tasks, it will wait for other tasks to be ready to begin to work. In the worst case, all jobs are pending leading to a deadlock. To solve this problem, co-scheduling is needed for the scheduler. The new scheduler framework makes the goal possible. 

Reviewer 3: s/different system/different controller

As you introduce fixes here and there, consider splitting paragraphs in more lines to ease the reviews.

Reviewer: nit: “Kubneretes”

 
## Goals
1. Use scheduler plugin, which is the most Kubernetes native way, to implement coscheduling. 
2. Lightweight implementation of coscheduling without the CRD of `PodGroup`
 
## Non-Goals
Use CRD `PodGroup` - this design proposes a lightweight approach that doesn't need to impose CRD management.
 
## Use Cases
When running a Tensorflow/MPI job, all tasks must start before they can do any work. This becomes a bigger problem when several jobs are competing to get all their tasks started. In worst case, all jobs are pending because of a deadlock: every job only start part of tasks, and waits for the other tasks to start. In worst case, all jobs are pending leading to a deadlock.

## Terms

- **pgPod**: pod belongs to some `PodGroup`.
- **regularPod**: a regular `Pod` (which doesn't have `PodGropuName` set).

## Proposal

> In order to implement coscheduling, we developed plugins in different extension points. In `QueueSort`  we ensure that the Pods belonging to the same PodGroup are queued back-to-back. For example, suppose PodGroup A owns Pod-A1, Pod-A2, Pod-A3, while PodGroup B owns Pod-B1, Pod-B2. The pods of the two PodGroups should not interleave - it should be always <PodGroup-A, PodGroup-B> or the other way around; but never <Pod-A1, Pod-B1, Pod-A2, ...>. In `Permit` phase we put the pod that doesn't meet min-available into the WaitingMap and reserve resources until min-available are met or timeout is triggered. In `Unreserve` phase，clean up the pods that timed-out.

Reviewer 3: nit: Too much detail for QueueSort in this summary.

Also, focus in the general idea. IIRC, it is only one pod group is processed at a time?

```yaml 
labels:
     pod-group.scheduling.sigs.k8s.io/name: nginx
     pod-group.scheduling.sigs.k8s.io/min-available: "2"
``` 
`Pods` in the same `PodGroup` with different priorities might lead to unintended behavior, so need to ensure `Pods` in the same `PodGroup` with the same priority.

### Coscheduling
```go
// Coscheduling is a plugin that implements the mechanism of gang scheduling.
type Coscheduling struct {
    FrameworkHandle     	framework.FrameworkHandle
    PodLister               corelisters.PodLister
    // Key is the name of PodGroup.
    PodGroupInfos       	map[string]PodGroupInfo
    // Name of the last scheduled podgroup
    LastPodGroup        	string
}
 
type PodGroupInfo struct {
    // LastFailureTimestamp stores the timestamp of last scheduling failure.
    LastFailureTimestamp 	time.Time
    UID           			types.UID
    MinAvailable  		int
    Name                                 String
}
```

> 1.  `PodGroupInfo` is initialized the first time the pod belongs to the PodGroup is encountered, and LastFailureTimestamp is updated every time the PodGroup fails to schedule.

Reviewer 3: again, what is "fail to schedule"?

Author: `fails to schedule` means that the `PodGroup` fails in 2 places
1. `Waiting time timeout` in `WaitingMap`
2.  the previous `PodGroup` is rejected by the second `PodGroup` in `PreFilter`

I just want to define a timestamp to compare the timestamp of `PodInfo` in `QueueSort`. But I find if I update the `LastFailureTimestamp`, it maybe lead lead to undefined behavior in the heap. So I change it to `InitialTimestamp` (InitialTimestamp stores the timestamp of the initialization time of PodGroup.). And compare the `InitialAttemptTimestamp` of `PodInfo`

Reviewer 3: sg, please add these notes to the KEP.

Author: thanks~~ I update in the KEP @alculquicondor

> 2.  `LastPodGroup` records which `PodGroup` the last scheduled pod belongs to.

Reviewer 3: scheduled, processed or reserved?

3.  `UID` is the unique identification value used to distinguish different podgroups.
 

### Extension points

#### QueueSort

> In order to make the pods which belongs to the same `PodGroup` to be scheduled together as much as possible, implement a strategy in `QueueSort` phase. 

Reviewer 3: We can't "ensure". We are only maximizing the chance.

> **limition**: `QueueSort` is the core part of our design and only one `QueueSort` plugin is allowed in the scheduling framework. So our design only supports the case that `QueueSort` extension point isn't implemented in other plugins.

Reviewer 3: s/limition/limitation

   
2. When podA and podB are both regularPods (we will check it by their labels), it follows the same logic of default in-tree [PrioritySort](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L41-L45) plugin.
   
3. When podA is regularPod, podB is pgPod, trying to order by podA's `Timestamp` (the time pod added to the scheduling queue) and podB’s `LastFailureTimestamp` (we get the PodGroupInfo from the cache). Pod with earlier timestamp is scheduled ahead of other pod.   
   
4. When podA and podB are both pgPods. 
   1. Trying to order by the `LastFailureTimestamp` of the podGroup. Pod with earlier timestamp is scheduled ahead of other pod.
   
   2. If `LastFailureTimestampA` is equal to `LastFailureTimestampB`, trying to order by the `UID` of `PodGroup`,`Pod` with lexicographically greater `UID` is scheduled ahead of other pod. (The purpose is to distinguish different `PodGroup` with the same `LastFailureTimestamp` and to keep the pods of the same `PodGroup` together)

> 

Reviewer 3: Just to emphasize what is different in this writing: it focuses on intention, rather than implementation details. This makes the design easier to understand.

**Note1**: There are different `LastFailureTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.

#### Pre-Filter
1. `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.

> 2. `PreFilter` validates that if we should reject `WaitingPods` in advance. When the pod belongs to a new `PodGroup` (we can check it by the `LastPodGroup` in cache), it indicates that the new PodGroup scheduling cycle has been entered. Then we will check if the last `PodGroup` meets the minAvailable, if not, we will reject it directly in advance. The purpose is 1.To avoid invalid waiting, 2.To make the pod belongs to the same `PodGroup` fail together, rather than waiting partially.

Reviewer 4: Can you show me why rejecting on-going waiting PodGroup is mandatory? It seems that we should use a Set (instead of String) for on-going PodGroup as a new PodGroup's scheduling is independent on existing on-going one, isn't it?

If we keep rejecting existing on-going PodGroup, it can still lead to deadlock even if right now we put Permit at the end of scheduling cycle.

Author: > Can you show me why rejecting on-going waiting PodGroup is mandatory?

The purpose of this is to make the pods belonging to the same `PodGroup` adjacent to each other in the process, rather than being queued by other pods.  eg: podA1, podA2, podA3(PodGroupA min=3),  podB1, podB2, podB3(PodGroupB min=3)

1. PodA1 PodA2  failed (Insufficient resources in the current cluster)
2. PodA3 success to WaitingMap (Can't actually schedule successfully because min=3)
3. PodB1 PodB2 failed (Insufficient resources in the current cluster)
4. PodB3 success to WaitingMap
5. Due to timeout mechanism, PodA3 and PodB3 are in WaitingMap. The new scheduling order is PodA1 PodA2 PodB1 PodB2. So PodGroupA is queued by other pods.

The above logic is to reduce the number of scheduling failures. But If it will lead to deadlock because rejecting existing on-going PodGroup, I think it is not necessary. WDYT? @Huang-Wei

Reviewer 4: Ah, I see the point, any Pod in a PodGroup cannot "reject" the on-going PodGroup - because you don't know if you're the last one and hence cannot tell whether reject or not. What a Pod in an on-going PodGroup can only "approve" - when the waitingPod number exceeds the minAvailable.

So the 2nd paragraph is more an optimization to fail fast the previous PodGroup - which is destined unable to be scheduled, please highlight that. And I think it's not necessarily only the (1st) Pod of another PodGroup can make this (reject) decision - any Pod other than the Pod belonging to the previous PodGroup can make the decision, right?

Reviewer 3: So you are basically trying to hold a second PodGroup until the first PodGroup has satisfied minAvailable. I have the feeling that Wei has a different interpretation.

But then pods from the second group will move to the unschedulable queue. How do you plan to bring them back once the first group satisfies it's lower limit?

Author: > And I think it's not necessarily only the (1st) Pod of another PodGroup can make this (reject) decision - any Pod other than the Pod belonging to the previous PodGroup can make the decision, right?

Yes, that's right. It is just optimized mechanism to fail fast the previous PodGroup. But I think it is difficult to chose when to reject the previous PodGroup. I temporarily removed this mechanism from the KEP.

> So you are basically trying to hold a second PodGroup until the first PodGroup has satisfied minAvailable. 

Not to hold the second PodGroup until the first PodGroup has satisfied minAvailable. It is to fail the first PodGroup fast if it is determined the first PodGroup is unable to satisfy minAvailable competition because the second PodGroup is coming is schedule cycle.

Reviewer 3: This is definitely not clear at all. Let's add the conditions for considering a pod group unsatisfiable.

But also I  don't think this is a good idea. Let's say some PG1 and PG2 (both requiring 3 min) are being created at almost the same time and the queue looks like this:

```
queue              | events about to come from apiserver
PG1, PG1, PG2, PG2 | PG1, PG2
```

The pods after the line are not in the queue yet, because there might be a delay from apiserver. But it seems that in your proposal, once the first PG2 is seen, we already reject PG1, because it's not satisfiable. But it will be soon.

Author: Yes,  you are right. There's something wrong with this mechanism, because it is difficult to chose when to reject the previous PodGroup. I remove this mechanism in `PreFilter`. Thank you @alculquicondor

---
title: Coscheduling plugin based on scheduler framework
authors:
  - "@denkensk"
owning-sig: sig-scheduling
reviewers:
  - "@Huang-Wei"
  - "@resouer"
  - "@hex108"
approvers:
  - "@Huang-Wei"
creation-date: 2020-01-16
last-updated: 2020-01-16
status: provisional
---

# Coscheduling plugin based on scheduler framework

## Table of Contents
<!-- toc -->
- [Motivation](#motivation)
- [Goals](#goals)
- [Non-Goals](#non-goals)
- [Use Cases](#use-cases)
- [Terms](#terms)
- [Proposal](#proposal)
- [Design Details](#design-details)
  - [PodGroup](#podgroup)
  - [Coscheduling](#coscheduling)
  - [Extension points](#extension-points)
    - [QueueSort](#queuesort)
    - [Pre-Filter](#pre-filter)
    - [Permit](#permit)
    - [UnReserve](#unreserve)
- [Alternatives considered](#alternatives-considered)
- [Graduation Criteria](#graduation-criteria)
- [Testing Plan](#testing-plan)
- [Implementation History](#implementation-history)
- [References](#references)
<!-- /toc -->

## Motivation

> Kubernetes has become a popular solution for orchestrating containerized workloads. Due to limitation of Kubernetes scheduler, some offline workloads(ML/DL)  are managed in a different system. For the sake of operation efficiency, we'd like to treat Kubneretes as a unified management platform. But the ML jobs have the characteristics of All-or-Nothing in the scheduling process, which requires all tasks of a job to be scheduled at the same time. If the job only start part of tasks, it will wait for other tasks be ready to begin to work. In worst case, all jobs are pending like deadlock. To solve the above problem, co-scheduling is needed for the scheduler. The new scheduler framework makes the goal possible.

Reviewer 4: ... for the sake of **improving cluster utilization** and operation efficiency...

Author: Done. thx

 
## Goals
 Use scheduler plugin, which is the most Kubernetes native way, to implement coscheduling.
 
## Non-Goals
 Discuss the API definition of `PodGroup`.
 
## Use Cases
When running a Tensorflow/MPI job, all tasks of a job must be start together; otherwise, did not start anyone of tasks. If the job only start part of tasks, it will wait for other tasks be ready to start work. In worst case, all jobs are pending like deadlock.
 
## Terms

- **pgPod**: pod belongs to some `PodGroup`.
- **regularPod**: a regular `Pod` (which doesn't have `PodGropuName` set).

## Proposal

> In order to implement the coscheduling, we developed plugins in different extension points. In `Sort` phase, add strategy to try to make sure that the pods belong to same PodGroup are queued continuously. In `Permit` phase，put the pod that doesn't meet min-available into WaitingMap and reserve resources until min-available are met or timeout. In `Unreserve` phase，clean up the pods which is timeout.

Reviewer 4: How about: In `QueueSort` phase, implement a strategy to ensure the Pods belonging to the same PodGroup are queued continously. For example, suppose PodGroup A owns Pod-A1, Pod-A2, Pod-A3, while PodGroup B owns Pod-B1, Pod-B2. The pods of the two PodGroups cannot be zigzagged - it should be always <PodGroup-A, PodGroup-B> or the other way around; it should NOT be any way like <Pod-A1, Pod-B1, Pod-A2, ...>.

Author: Done thx

> ![image](20200116-coscheduling-plugin-based-on-scheduler-framework-extensions.png)

Reviewer 4: Can you put it somewhere and comment with the URL? so that it can be viewed during the review.

Author: Done. thx.  I add a comment

```yaml 
labels:
     pod-group/name: nginx
     pod-group/min-available: "2"
``` 

### Coscheduling
```go
// Coscheduling is a plugin that implements the mechanism of gang scheduling.
type Coscheduling struct {
    FrameworkHandle     	framework.FrameworkHandle
    PodLister               corelisters.PodLister
    // Key is the name of PodGroup.
    PodGroupInfos       	map[string]PodGroupInfo
    // Name of the last scheduled podgroup
    LastPodGroup        	string
}
 
type PodGroupInfo struct {
    // LastFailureTimestamp stores the timestamp of last scheduling failure.
    LastFailureTimestamp 	time.Time
    UID           			types.UID
    MinAvailable  		int
    Name                                 String
}
```

1.  `PodGroupInfo` is initialized the first time the pod belongs to the PodGroup is encountered, and LastFailureTimestamp is updated every time the PodGroup fails to schedule.
2.  `LastPodGroup` records which `PodGroup` the last scheduled pod belongs to.
3.  `UID` is the unique identification value used to distinguish different podgroups.
 

### Extension points

#### QueueSort

> In order to make the pods which belongs to the same `PodGroup` to be scheduled together as much as possible, much policies are added in `QueueSort` phase. 

Reviewer 4: What do you mean by "much policies"?

Author: I change it to `implement a strategy in `QueueSort` phase`.

```go
  func  Less(podA *PodInfo, podB *PodInfo) bool
```

> 1. When priorities are different, they are compared based on their priorities. When priorities are the same, they are operated according to the following process.

Reviewer 4: When priorities (i.e. `.spec.priorityValue`) are different

Author: Done thx

> 2. When podA and podB are both regularPods (we will check it by the label of pod), it follows the same logic of default in-tree `PrioritySort` plugin.

Reviewer 4: ... we will check it by their labels ...

Reviewer 4: Add a link for [plugin](https://github.com/kubernetes/kubernetes/blob/f88f58cb7938ee54216b390b07dd0745ee59fcc1/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L41-L45).

Author: Done thx

---
title: Coscheduling plugin based on scheduler framework
authors:
  - "@denkensk"
owning-sig: sig-scheduling
reviewers:
  - "@Huang-Wei"
  - "@resouer"
  - "@hex108"
approvers:
  - "@Huang-Wei"
creation-date: 2020-01-16
last-updated: 2020-01-16
status: provisional
---

# Coscheduling plugin based on scheduler framework

## Table of Contents

<!-- toc -->
- [Motivation](#motivation)
- [Goals](#goals)
- [Non-Goals](#non-goals)
- [Use Cases](#use-cases)
- [Terms](#terms)
- [Proposal](#proposal)
- [Design Details](#design-details)
  - [PodGroup](#podgroup)
  - [Coscheduling](#coscheduling)
  - [Extension points](#extension-points)
    - [QueueSort](#queuesort)
    - [Pre-Filter](#pre-filter)
    - [Permit](#permit)
    - [UnReserve](#unreserve)
- [Alternatives considered](#alternatives-considered)
- [Graduation Criteria](#graduation-criteria)
- [Testing Plan](#testing-plan)
- [Implementation History](#implementation-history)
- [References](#references)
<!-- /toc -->

## Motivation

> Kubernetes has become a popular solution for orchestrating containerized workloads. Due to limitation of Kubernetes scheduler, some offline workloads(ML/DL)  are managed in a different system. For the sake of improving cluster utilization and operation efficiency., we'd like to treat Kubneretes as a unified management platform. But the ML jobs have the characteristics of All-or-Nothing in the scheduling process, which requires all tasks of a job to be scheduled at the same time. If the job only start part of tasks, it will wait for other tasks be ready to begin to work. In worst case, all jobs are pending like deadlock. To solve the above problem, co-scheduling is needed for the scheduler. The new scheduler framework makes the goal possible.

Reviewer 5: ```suggestion
Kubernetes has become a popular solution for orchestrating containerized workloads. Due to limitation of Kubernetes scheduler, some offline workloads (ML/DL) are managed in a different system. To improve cluster utilization and operation efficiency, we'd like to treat Kubneretes as a unified management platform. But ML jobs are All-or-Nothing: they require all tasks of a job to be scheduled at the same time. If the job only start part of tasks, it will wait for other tasks to be ready to begin to work. In the worst case, all jobs are pending leading to a deadlock. To solve this problem, co-scheduling is needed for the scheduler. The new scheduler framework makes the goal possible.
```

 
## Goals
 Use scheduler plugin, which is the most Kubernetes native way, to implement coscheduling.
 
## Non-Goals
 Discuss the API definition of `PodGroup`.
 
## Use Cases

> When running a Tensorflow/MPI job, all tasks of a job must be start together; otherwise, did not start anyone of tasks. If the job only start part of tasks, it will wait for other tasks be ready to start work. In worst case, all jobs are pending like deadlock.

Reviewer 5: ```suggestion
When running a Tensorflow/MPI job, all tasks of a job must start or none. If the job only start part of tasks, it will wait for other tasks to be ready to start work. In worst case, all jobs are pending leading to a deadlock.
```

Reviewer 5: when you say "it will wait for other tasks", who exactly waits? the tasks themselves keep waiting? or are you talking conceptually that the tasks should by some mechanism keep waiting (and so in our scheduler this is achieve by waiting at Permit)?

Author: The tasks themselves keep waiting, when the job only start part of tasks, Started tasks will for not started tasks to be ready to begin to work. This will lead to waste of resources. So the tasks should by some mechanism keep waiting in scheduler, so in our scheduler this is achieve by waiting at Permit.

 
## Terms

- **pgPod**: pod belongs to some `PodGroup`.
- **regularPod**: a regular `Pod` (which doesn't have `PodGropuName` set).

## Proposal

> In order to implement the coscheduling, we developed plugins in different extension points. In `QueueSort` phase, implement a strategy to ensure the Pods belonging to the same PodGroup are queued continuously. For example, suppose PodGroup A owns Pod-A1, Pod-A2, Pod-A3, while PodGroup B owns Pod-B1, Pod-B2. The pods of the two PodGroups cannot be zigzagged - it should be always <PodGroup-A, PodGroup-B> or the other way around; it should NOT be any way like <Pod-A1, Pod-B1, Pod-A2, ...>In `Permit` phase，put the pod that doesn't meet min-available into WaitingMap and reserve resources until min-available are met or timeout. In `Unreserve` phase，clean up the pods which is timeout.

Reviewer 5: ```suggestion
In order to implement coscheduling, we developed plugins in different extension points. In `QueueSort`  we ensure that the Pods belonging to the same PodGroup are queued back-to-back. For example, suppose PodGroup A owns Pod-A1, Pod-A2, Pod-A3, while PodGroup B owns Pod-B1, Pod-B2. The pods of the two PodGroups should not interleave - it should be always <PodGroup-A, PodGroup-B> or the other way around; but never <Pod-A1, Pod-B1, Pod-A2, ...>. In `Permit` phase we put the pod that doesn't meet min-available into the WaitingMap and reserve resources until min-available are met or timeout is triggered. In `Unreserve` phase，clean up the pods that timed-out.
```

![image](./20200116-coscheduling-plugin-based-on-scheduler-framework-extensions.png)

## Design Details

### PodGroup

>  `Pod` is distinguished by its `labels`. Pod of the same `PodGroup` has the same name. We use the label to distinguish the `PodGroup` for less dependence on the api definition of `PodGroup`.  It's just a simple description of the `PodGroup`. If the definition of `PodGroup` is accepted by the community, we will define it directly through the CRD of `PodGroup`. This is not the focus of this proposal.

Reviewer 5: what do you mean by ```pod of the same PodGroup has the same name```? do you mean the prefix?

Reviewer 5: oh, you mean the podgroup name is the same.

>  `Pod` is distinguished by its `labels`. Pod of the same `PodGroup` has the same name. We use the label to distinguish the `PodGroup` for less dependence on the api definition of `PodGroup`.  It's just a simple description of the `PodGroup`. If the definition of `PodGroup` is accepted by the community, we will define it directly through the CRD of `PodGroup`. This is not the focus of this proposal.

Reviewer 5: ```suggestion
We use a special label named ```pod-group/name``` to define a PodGroup. Pods that set this label and use the same value belong to the same PodGroup. This is a short term solution, in the future if the definition of `PodGroup` is accepted by the community, we will define it directly through the CRD of `PodGroup`. This is not the focus of this proposal.
```

2.  `LastPodGroup` records which `PodGroup` the last scheduled pod belongs to.
3.  `UID` is the unique identification value used to distinguish different podgroups.
 

### Extension points

#### QueueSort
In order to make the pods which belongs to the same `PodGroup` to be scheduled together as much as possible, implement a strategy in `QueueSort` phase. 

```go
  func  Less(podA *PodInfo, podB *PodInfo) bool
```
1. When priorities (i.e. .spec.priorityValue) are different, they are compared based on their priorities. When priorities are the same, they are operated according to the following process.
   
2. When podA and podB are both regularPods (we will check it by their labels), it follows the same logic of default in-tree [PrioritySort](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L41-L45) plugin.
   
3. When podA is regularPod, podB is pgPod. they are compared according to podA’s timestamp and podB’s `LastFailureTimestamp` (we get the PodGroupInfo from the cache). If timestampA is earlier than LastFailureTimestampB, return true. Otherwise, return false. When podA is pgPod, podB is regularPod. Ditto
   

> 4. When podA and podB are both pgPods. Compare the the `LastFailureTimestamp` of the podGroup. 

Reviewer 5: similar to the comment above, instead of describing implementation, describe the goal: what order is this step trying to achieve.

   1. If `LastFailureTimestampA` is earlier than `LastFailureTimestampB`, return true. `LastFailureTimestampA` is later than `LastFailureTimestampB`, return false.
   
   2. If `LastFailureTimestampA` is equal to `LastFailureTimestampB`, we will compare the `UID` of `PodGroup`. (The purpose is to distinguish different `PodGroup` with the same `LastFailureTimestamp` and to keep the pods of the same `PodGroup` together)

**Note1**: There are different `LastFailureTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.

#### Pre-Filter

> 1. When a pod which belongs to a `PodGroup` comes, calculate the total number of Pods  excludes terminating ones belongs to the same PodGroup. If the result is less than minAvailable, the scheduling process is terminated directly.

Reviewer 5: ```suggestion
1. When a pod which belongs to a `PodGroup` comes, calculate the total number of Pods that belong to the same PodGroup excluding terminating ones. If the result is less than minAvailable, the scheduling process is terminated directly.
```

Reviewer 5: as in PreFilter returns unschedulable? note that this will cause a scheduling failure not a "fit error": see here: https://github.com/kubernetes/kubernetes/blob/e6b5194ec12a7aa241c702019d6cd6f360b232d5/pkg/scheduler/core/generic_scheduler.go#L187

Reviewer 5: I am not sure I understand what you are tying to achieve, so similar to the comments above, start by describing the goal that PreFilter is trying to achieve not its implementation. For example: "PreFilter validates that ...."

Author: `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly. 
I change it to return `UnschedulableAndUnresolvable` in PreFilter here.

Reviewer 5: It doesn't matter whether the return status is ```UnschedulableAndUnresolvable``` or ```Unschedulable```, in both cases the scheduler will return a normal error, not a fit error, look at the code I pointed to in the comment above.

Author: em.. @ahg-g  I think it is not a fit error, maybe a normal error.  For example, the current total number of all pods belongs to a same `PodGroup` is 3 (Maybe there are other reasons why the left pods haven't been created yet), but the `minAvailable` of the `PodGroup` is 5. I think it is better to abort the scheduling cycle  in `PreFilter `.

Reviewer 4: I think it's meant to fail quickly for the first N-1 unassigned Pods, while recording the number in the cache. Then when the Nth Pod comes, scheduler knows there are N-1 Pods pending in the internal queue, so it proceeds the Nth Pod with the reguarlar scheduling cycle. Is it right? @denkensk

```go
func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
    …
 
     // Count the numbers of all pods belongs to the same PodGroup excludes terminating ones.
    total := cs.calculateTotalPods(podGroupName)
    if total < minAvailable {
        return framework.NewStatus(framework.Unschedulable, "less than minAvailable")
    }
    ...
    if cs.Last != "" && cs.Last != podGroupName {
        allocated := cs.calculateAllocatedPods(cs.LastPodGroup)
        pgInfo, ok := cs.PodGroupInfos[cs.LastPodGroup]
        if ok {
        	pgInfo.LastFailureTimestamp = time.Now()
        	if allocated < pgInfo.minAvailable {
            	cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
                	if p.GetPod().Labels[PodGroupName] == cs.LastPodGroup {
                    	p.Reject(cs.Name())
                	}
            	})
        	}
        }
    }
    cs.LastPodGroup = podGroupName
    return nil
}
```

#### Permit
1. Get the number of Running pods that belong to the same PodGroup
2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup

> 3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), apprve the waiting pods that  belongs to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout(eg: 10s)

Reviewer 5: ```suggestion
3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: 10s).
```

 
```go
func (cs Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
    ...
    running := cs.calculateRunningPods(podGroupName)
    waiting := cs.calculateWaitingPods(podGroupName)
    if running+waiting+1 < minAvailable {
        return framework.NewStatus(framework.Wait, ""), 10 * time.Second
    }
    ...
    cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
        if p.GetPod().Labels[PodGroupName] == podGroupName {
        	p.Allow(cs.Name())
        }
    })
   
    delete(cs.PodGroupInfos, podGroupName)
    return framework.NewStatus(framework.Success, ""), 0
}
```
 

#### UnReserve

> After a pod which belongs to a PodGroup times out in the permit phase.  We will release the pods belongs to the same PodGroup to avoid long-term invalid occupy of resources.

Reviewer 5: ```suggestion
After a pod which belongs to a PodGroup times out in the permit phase.  UnReserve ```Rejects``` the pods that belong to the same PodGroup to avoid long-term invalid reservation of resources.
```

---
title: Coscheduling plugin based on scheduler framework
authors:
  - "@denkensk"
owning-sig: sig-scheduling
reviewers:
  - "@Huang-Wei"
  - "@resouer"

>   - "@hex108"

Reviewer 6: Could you kindly add me to reviewers??

Author: Done thanks ~~

approvers:
  - "@Huang-Wei"
creation-date: 2020-01-16
last-updated: 2020-01-16
status: provisional
---

# Coscheduling plugin based on scheduler framework

## Table of Contents

<!-- toc -->
- [Motivation](#motivation)
- [Goals](#goals)
- [Non-Goals](#non-goals)
- [Use Cases](#use-cases)
- [Terms](#terms)
- [Proposal](#proposal)
- [Design Details](#design-details)
  - [PodGroup](#podgroup)
  - [Coscheduling](#coscheduling)
  - [Extension points](#extension-points)
    - [QueueSort](#queuesort)
    - [Pre-Filter](#pre-filter)
    - [Permit](#permit)
    - [UnReserve](#unreserve)
- [Alternatives considered](#alternatives-considered)
- [Graduation Criteria](#graduation-criteria)
- [Testing Plan](#testing-plan)
- [Implementation History](#implementation-history)
- [References](#references)
<!-- /toc -->

## Motivation

> Kubernetes has become a popular solution for orchestrating containerized workloads. Due to limitation of Kubernetes scheduler, some offline workloads (ML/DL) are managed in a different system. To improve cluster utilization and operation efficiency, we'd like to treat Kubneretes as a unified management platform. But ML jobs are All-or-Nothing: they require all tasks of a job to be scheduled at the same time. If the job only start part of tasks, it will wait for other tasks to be ready to begin to work. In the worst case, all jobs are pending leading to a deadlock. To solve this problem, co-scheduling is needed for the scheduler. The new scheduler framework makes the goal possible.

Reviewer 3: It's worth mentioning the specific limitations here. It's the "Motivation" section after all :)

 
## Goals
 Use scheduler plugin, which is the most Kubernetes native way, to implement coscheduling.
 
## Non-Goals

>  Discuss the API definition of `PodGroup`.

Reviewer 7: we have discussed PodGroup before, are we going to introduce some new one?

 
## Use Cases

> when running a Tensorflow/MPI job, all tasks of a job must be start together; otherwise, did not start anyone of tasks. If the resource is enough to run all 'tasks', everything is fine; but it's not true for most of case, especially in the on-premises environment. In worst case, all jobs are pending here because of deadlock: every job only start part of tasks, and waits for the other tasks to start. In worst case, all jobs are pending leading to a deadlock.

Reviewer 7: > It can happen in any environment.

In cloud environment, they may consider autoscale feature; but cost is another factor to user.

> when running a Tensorflow/MPI job, all tasks of a job must be start together; otherwise, did not start anyone of tasks. If the resource is enough to run all 'tasks', everything is fine; but it's not true for most of case, especially in the on-premises environment. In worst case, all jobs are pending here because of deadlock: every job only start part of tasks, and waits for the other tasks to start. In worst case, all jobs are pending leading to a deadlock.

Reviewer 3: Even if they do use autoscaling, most users will likely have upper limits.

When running a Tensorflow/MPI job, all tasks of a job must be start together; otherwise, did not start anyone of tasks. If the job only start part of tasks, it will wait for other tasks be ready to start work. In worst case, all jobs are pending like deadlock.
 
## Terms

- **pgPod**: pod belongs to some `PodGroup`.
- **regularPod**: a regular `Pod` (which doesn't have `PodGropuName` set).

## Proposal

In order to implement the coscheduling, we developed plugins in different extension points. In `QueueSort` phase, implement a strategy to ensure the Pods belonging to the same PodGroup are queued continously. For example, suppose PodGroup A owns Pod-A1, Pod-A2, Pod-A3, while PodGroup B owns Pod-B1, Pod-B2. The pods of the two PodGroups cannot be zigzagged - it should be always <PodGroup-A, PodGroup-B> or the other way around; it should NOT be any way like <Pod-A1, Pod-B1, Pod-A2, ...>In `Permit` phase，put the pod that doesn't meet min-available into WaitingMap and reserve resources until min-available are met or timeout. In `Unreserve` phase，clean up the pods which is timeout.

> ![image](./20200116-coscheduling-plugin-based-on-scheduler-framework-extensions.png)

Author: https://github.com/denkensk/enhancements/blob/coscheduling-plugin/keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework-extensions.png

> 

Author: https://github.com/denkensk/enhancements/blob/coscheduling-plugin/keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework-extensions.png

## Design Details

### PodGroup

> We use a special label named ```pod-group/name``` to define a PodGroup. Pods that set this label and use the same value belong to the same PodGroup. This is a short term solution, in the future if the definition of `PodGroup` is accepted by the community, we will define it directly through the CRD of `PodGroup`. This is not the focus of this proposal.

Author: Done thx

```yaml 
labels:
     pod-group/name: nginx
     pod-group/min-available: "2"
``` 

### Coscheduling
```go
// Coscheduling is a plugin that implements the mechanism of gang scheduling.

> type Coscheduling struct {

Reviewer 2: So why not use `Gang` as the plugin name?

    FrameworkHandle     	framework.FrameworkHandle
    PodLister               corelisters.PodLister
    // Key is the name of PodGroup.
    PodGroupInfos       	map[string]PodGroupInfo

>     // Name of the last scheduled podgroup

Author: yes

>     // Name of the last scheduled podgroup

Reviewer 2: if two PodGroups have the same name, how to process the `LastPodGroup`

>     // Name of the last scheduled podgroup

Reviewer 3: Please update the comment accordingly

    LastPodGroup        	string
}
 
type PodGroupInfo struct {
    // LastFailureTimestamp stores the timestamp of last scheduling failure.

>     LastFailureTimestamp 	time.Time

Reviewer 3: ping

>     LastFailureTimestamp 	time.Time

Author: `fails to schedule` means that the `PodGroup` fails in 2 places
1. `Waiting time timeout` in `WaitingMap`
2.  the previous `PodGroup` is rejected by the second `PodGroup` in `PreFilter`

I just want to define a timestamp to compare the timestamp of `PodInfo` in `QueueSort`. But I find if I update the `LastFailureTimestamp`, it maybe lead lead to undefined behavior in the heap. So I change it to `InitialTimestamp` (InitialTimestamp stores the timestamp of the initialization time of PodGroup.). And compare the `InitialAttemptTimestamp` of `PodInfo`

    UID           			types.UID
    MinAvailable  		int
    Name                                 String
}
```

1.  `PodGroupInfo` is initialized the first time the pod belongs to the PodGroup is encountered, and LastFailureTimestamp is updated every time the PodGroup fails to schedule.
2.  `LastPodGroup` records which `PodGroup` the last scheduled pod belongs to.
3.  `UID` is the unique identification value used to distinguish different podgroups.
 

### Extension points

> #### QueueSort

Reviewer 6: I'd like to discuss implementing `QueueSort` extension points in the Coscheduling plugin.  It is because __only one __ `QueueSort` plugin is allowed in the scheduling framework.    What if some user already has some plugin that implements `QueueSort` plugin? And,  I think we better to document how we support the case after discussion.

Reviewer 7: Yes, we should not dependent on timeout which impact the throughput of scheduler.

In order to make the pods which belongs to the same `PodGroup` to be scheduled together as much as possible, implement a strategy in `QueueSort` phase. 

```go
  func  Less(podA *PodInfo, podB *PodInfo) bool
```

> 1. Trying to order by pod priority (i.e. .spec.priorityValue), pod with higher priority is scheduled ahead of other pod with lower priority. When priorities are the same, they are operated according to the following process.

Reviewer 7: what will happen if scheduler get low/high priority pods intervally, depdent on timeout? That's a common case under high loads :)

Reviewer 3: Within the same podgroup or in general?

> 2. When podA and podB are both regularPods (we will check it by their labels), it follows the same logic of default in-tree [PrioritySort](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L41-L45) plugin.

Reviewer 7: what will happen if we have more pods than minAvailable? how to guarantee that?

Author: When the count of pods belongs to the same `PodGroup` successfully scheduled is greater than min-available, it will be allowed in `permit`.

> 3. When podA is regularPod, podB is pgPod, trying to order by podA's `Timestamp` (the time pod added to the scheduling queue) and podB’s `LastFailureTimestamp` (we get the PodGroupInfo from the cache). Pod with earlier timestamp is scheduled ahead of other pod.   

Author: > LastFailureTimestamp of the group, right?
Yes

> Is that the intention? It worth adding what the intention is, because sometimes it's hard to grasp from "inequalities".
The individual pod has an `Timestamp` in `PodInfo` to identify when its last schedule failed or the time pod added to the scheduling queue. `Timestamp` is used in default `Less`. But for a PodGroup, a pod belonging to the same PodGroup should have a time to indicate the failure time for the whole PodGroup. Having the same `LastFailureTimestamp` will help all pods to be adjacent in the queue.

> Also, when LastFailureTimestamp changes, it is implicitly changing the heap priority of several pods. I think this could lead to undefined behavior in the heap.
Yes, this maybe lead to undefined behavior in the heap.  I think I can change the `LastFailureTimestamp` to `InitialTimestamp`. It will not change.  And also in regularPod.
WDYT?
@Huang-Wei @alculquicondor

> 3. When podA is regularPod, podB is pgPod, trying to order by podA's `Timestamp` (the time pod added to the scheduling queue) and podB’s `LastFailureTimestamp` (we get the PodGroupInfo from the cache). Pod with earlier timestamp is scheduled ahead of other pod.   

Reviewer 3: Define `LastFailureTimestamp` and `InitialTimestamp`.

> 4. When podA and podB are both pgPods.

Reviewer 6: Question: How are`priority` values of pgPods handled in this case??

Reviewer 6: I noticed that you clearly stated in step `1.`

Author: Done. thanks

   1. Trying to order by the `LastFailureTimestamp` of the podGroup. Pod with earlier timestamp is scheduled ahead of other pod.
   
   2. If `LastFailureTimestampA` is equal to `LastFailureTimestampB`, trying to order by the `UID` of `PodGroup`,`Pod` with lexicographically greater `UID` is scheduled ahead of other pod. (The purpose is to distinguish different `PodGroup` with the same `LastFailureTimestamp` and to keep the pods of the same `PodGroup` together)

**Note1**: There are different `LastFailureTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.

#### Pre-Filter

> 1. `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.

Reviewer 3: Is this trying to make sure all pods that belong to a podgroup are already in the queues?

Also, what is "interrupted"? I assume this means that it would return `UnschedulableAndUnresolvable`? What happens with `LastFailureTimestamp`?

Reviewer 3: Have we given up on the idea of only allowing one podgroup to be scheduled at a time? This design might still lead to deadlocks if more than one pod group already has minpods in the queues.

Reviewer 5: sorting the pods by their group id in the queue tries to approximate scheduling one pod group at a time, but this approach indeed does not offer any guarantees.

Reviewer 4: > This design might still lead to deadlocks if more than one pod group already has minpods in the queues.

I think so. So I'm questioning why it only allows ONE on-going PodGroup. See my comment below.

> 2. `PreFilter` validates that if we should reject `WaitingPods` in advance. When the pod belongs to a new `PodGroup` (we can check it by the `LastPodGroup` in cache), it indicates that the new PodGroup scheduling cycle has been entered. Then we will check if the last `PodGroup` meets the minAvailable, if not, we will reject it directly in advance. The purpose is 1. To avoid invalid waiting, 2. To make the pod belongs to the same `PodGroup` fail together, rather than waiting partially.

Reviewer 3: Why are these paragraphs numbered? I assumed it meant different steps, but it doesn't look like it.

What happens when the limit is passed? We should be moving all the pods back to the active queue, right? I don't think we have a way to do this :disappointed:

Reviewer 8: Should we can check all available resources in the cluster? If not meet the minAvaliable requirement of Pods, reject any pod of the PodGroup.

Reviewer 3: That's tricky. It sounds like overlap with `NodeResourcesFit`. But then, what about other filters? They are equally important.

```go
func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
    ...
 
     // Count the numbers of all pods belongs to the same PodGroup excludes terminating ones.

>     total := cs.calculateTotalPods(podGroupName)

Reviewer 3: are we counting the incoming pod?

    if total < minAvailable {
        return framework.NewStatus(framework.UnschedulableAndUnresolvable, "less than minAvailable")
    }
    ...
    if cs.Last != "" && cs.Last != podGroupName {

>         allocated := cs.calculateAllocatedPods(cs.LastPodGroup)

Reviewer 2: which pods are allocated pods?

Reviewer 2: If a pod is allocated, but it is deleted for some reason,  how does the plugin handle these pods?

        pgInfo, ok := cs.PodGroupInfos[cs.LastPodGroup]
        if ok {
        	pgInfo.LastFailureTimestamp = time.Now()
        	if allocated < pgInfo.minAvailable {
            	cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
                	if p.GetPod().Labels[PodGroupName] == cs.LastPodGroup {
                    	p.Reject(cs.Name())
                	}
            	})
        	}

>         }

Reviewer 6: Yes we should handle mutex podgroups in this point.

ref: `Prefilter` section in this [design doc](https://docs.google.com/document/d/1-igzuDyiN24yvcc0I4qGnN_w6tYRd9IZpEOgxrC7j_U/edit#heading=h.cwj3gqddiyrt)

>         }

Reviewer 3: Why are we reviewing code in a KEP? :sweat_smile: 

@denkensk can you express this in words? Focus in the intent.

>         }

Author: Thanks, I maybe a deadlock. This situation has been solved after this pr is merged  https://github.com/kubernetes/kubernetes/pull/88199/files 
Scheduler framework permit plugins now run at the end of the scheduling cycle, after reserve plugins. Waiting on permit will remain in the beginning of the binding cycle.

    }
    cs.LastPodGroup = podGroupName
    return nil
}
```

#### Permit
`Permit` phase we put the pod that doesn't meet min-available into the WaitingMap and reserve resources until min-available are met or timeout is triggered.
1. Get the number of Running pods that belong to the same PodGroup
2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: 10s).
 
```go
func (cs Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
    ...
    running := cs.calculateRunningPods(podGroupName)

>     waiting := cs.calculateWaitingPods(podGroupName)

Reviewer 3: I would say remove the code instead. This one is already expressed in words. The previous one needs to be expressed.

    if running+waiting+1 < minAvailable {

>         return framework.NewStatus(framework.Wait, ""), 10 * time.Second

Reviewer 6: I think the timeout would be better to be configurable.

Reviewer 3: I think different priorities in the same pod group should be explicitly unsupported (we don't guarantee that it will work)

The timeout might also depend on the size of the group. So rather than absolute timeout, we could have a factor to multiply against the minsize.

Author: @alculquicondor  @everpeace thanks, I will change the absolute timeout to dynamic value depends on the size of the group.

    }
    ...
    cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
        if p.GetPod().Labels[PodGroupName] == podGroupName {
        	p.Allow(cs.Name())
        }
    })
   
    delete(cs.PodGroupInfos, podGroupName)
    return framework.NewStatus(framework.Success, ""), 0
}
```
 

#### UnReserve
After a pod which belongs to a PodGroup times out in the permit phase.  UnReserve ```Rejects``` the pods that belong to the same PodGroup to avoid long-term invalid reservation of resources.

```go 
func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) {
    ...
    cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
        if p.GetPod().Labels[PodGroupName] == podGroupName {
        	p.Reject(cs.Name())
        }
    })
}
```

## Alternatives considered

> 1. Change the unit of Queue from `Pod` to `PodGroup`.

Reviewer 3: I assume this refers to using `PodGroup` as a scheduling unit.

The downside is badly expressed. It should be: this requires major refactoring, which only supports Pods as scheduling unit today.

            	})
        	}
        }
    }
    cs.LastPodGroup = podGroupName
    return nil
}
```

#### Permit
In `Permit` phase, we put the pod that doesn't meet min-available into the WaitingMap and reserve resources until min-available are met or timeout is triggered.
1. Get the number of Running pods that belong to the same PodGroup
2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: 10s).
 
```go
func (cs Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
    ...
    running := cs.calculateRunningPods(podGroupName)
    waiting := cs.calculateWaitingPods(podGroupName)
    if running+waiting+1 < minAvailable {
        return framework.NewStatus(framework.Wait, ""), 10 * time.Second
    }
    ...
    cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
        if p.GetPod().Labels[PodGroupName] == podGroupName {

>         	p.Allow(cs.Name())

Reviewer 9: Must `PodGroupName` be unique?

        }
    })
   
    delete(cs.PodGroupInfos, podGroupName)
    return framework.NewStatus(framework.Success, ""), 0
}
```
 

#### UnReserve
After a pod which belongs to a PodGroup times out in the permit phase.  UnReserve ```Rejects``` the pods that belong to the same PodGroup to avoid long-term invalid reservation of resources.

```go 
func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) {
    ...
    cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
        if p.GetPod().Labels[PodGroupName] == podGroupName {
        	p.Reject(cs.Name())
        }
    })
}
```

## Alternatives considered
1. Using `PodGroup` as a scheduling unit. This requires major refactoring, which only supports Pods as scheduling unit today.

> 

Reviewer 9: `Risks and Mitigations `are also needed, for example the current scheduler's performance is at 100/s. If `PodGroup` is enabled, then some measures are needed to ease the pressure on the scheduler.

    PodLister               corelisters.PodLister
    // Key is the name of PodGroup.
    PodGroupInfos       	map[string]PodGroupInfo
}
 
type PodGroupInfo struct {
    // InitialTimestamp stores the timestamp of the initialization time of PodGroup.
    InitialTimestamp 	time.Time
    UID           			types.UID
    MinAvailable  		int
    Name                                 String
}
```

> 1.  `PodGroupInfo` is initialized the first time the pod belongs to the PodGroup is encountered, and LastFailureTimestamp is updated every time the PodGroup fails to schedule.

Reviewer 4: Please remove `and LastFailureTimestamp is updated every time the PodGroup fails to schedule.`.

> 1.  `PodGroupInfo` is initialized the first time the pod belongs to the PodGroup is encountered, and LastFailureTimestamp is updated every time the PodGroup fails to schedule.

Author: Done

#### QueueSort
In order to maximize the chance that the pods which belong to the same `PodGroup` to be scheduled consecutively, we need to implement a customized `QueueSort` plugin to sort the Pods properly.

**limitation**: `QueueSort` is the core part of our design and only one `QueueSort` plugin is allowed in the scheduling framework. So our design only supports the case that `QueueSort` extension point isn't implemented in other plugins.

```go
func  Less(podA *PodInfo, podB *PodInfo) bool
```

Firstly, we will inherit the default in-tree PrioritySort plugin so as to honor .spec.priority to ensure high-priority Pods are always sorted ahead of low-priority ones.

Secondly, if two Pods hold the same priority, the sorting precedence is described as below:

- If they are both regularPods (without particular PodGroup label), compare their `InitialAttemptTimestamp` field: the Pod with earlier `InitialAttemptTimestamp` is positioned ahead of the other.
  
- If one is regularPod and the other is pgPod, compare regularPod's `InitialAttemptTimestamp` with pgPod's `InitialTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
  
- If they are both pgPods:

>   - Compare their `InitialAttemptTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.

Reviewer 3: Have you thought of starvation?
If two pod groups have the same priority, the first one to enter the queue will always have higher precedence than the other, or even individual pods.

This is different than the original, where the scheduler tries to give the same scheduling opportunities to all pods. But, on the other hand, this might be something that we want for co-scheduling.

Ref https://github.com/kubernetes/kubernetes/issues/83834#issuecomment-587557978
cc @yqwang-ms

Reviewer 10: Thanks for cc.

BTW, seems co-scheduling by itself just needs "Gang Queuing", i.e.:
> Pods belonging to the same PodGroup are queued back-to-back

Which is not necessary to be also FIFO. (And FIFO might NOT be something that we want for co-scheduling, it is more like that to implement co-scheduling, to be FIFO is a helper, so we coupled them)

Given current K8S has priority starvation issues mentioned in https://github.com/kubernetes/kubernetes/issues/86373.
E.g. an early large scale pod group, keeps in waiting, will also cause later small pod group to starve, even if the small pod group can fit into free resource, and from user view, they have the same priority, and just only a little different creation time.

So, we would better decouple FIFO with co-scheduling, such as, for sometime, we move the whole early group behind later group, and for other times, we move the whole back.

Reviewer 4: > If two pod groups have the same priority, the first one to enter the queue will always have higher precedence than the other, or even individual pods.

Technically, due to the limitation of the current global one-queue mechanism, it's possible upon a large pod group as it's pretty challenging to accommodate all precedence into one queue.

> But, on the other hand, this might be something that we want for co-scheduling.

Comparing to the way that randomly sorts PodGroup Pods (where wait() is more unpredictable), this approach seems to have more chances to avoid starvation. However, as I mentioned above, technically it still has chances.

Reviewer 3: > So, we would better decouple FIFO with co-scheduling, such as, for sometime, we move the whole early group behind later group, and for other times, we move the whole back.

Could you rephrase this? Did you mean couple or decouple? What would be "other times"?

Reviewer 10: I mean we would better decouple them given current situation.
For example, a large early pod group A, and another small later pod group B.
For queuing order, could we make A has 70% chance ahead of B? (So B still has 30% chance ahead of A)? 
So in this way, we at least can mitigate or avoid the starvation for pods of the same priority. Otherwise, same priority pods also suffers from this proposal, right?

Reviewer 10: I mean can we easily queue pods in a group as a gang/unit, instead of leveraging detailed tuned compare rules to make sure they are back-to-back (has side effects to be FIFO)?

Author: > I mean can we easily queue pods in a group as a gang/unit, instead of leveraging detailed tuned compare rules to make sure they are back-to-back (has side effects to be FIFO)?

I mentioned it in `Alternatives considered`. I think this requires major refactoring, which only supports Pods as scheduling unit today.  @yqwang-ms

  - If their `InitialAttemptTimestamp` is identical, order by their UID of PodGroup: a Pod with lexicographically greater UID is scheduled ahead of the other Pod. (The purpose is to tease different PodGroups with the same `InitialAttemptTimestamp` apart, while also keeping Pods belonging to the same PodGroup back-to-back)

> **Note1**: There are different `InitialTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.

Author: Done

#### Pre-Filter
When a `pgGroup` Pod is being scheduled for the first time, we have 2 option to deal with it: either start the full scheduling cycle no matter its grouping Pods are present inside schedule queue, or fail quick its scheduling cycle as its grouping Pods number doesn't meet `minAvailable`. The former case is more efficient, but may cause partial Pods holding system resources until a timeout. The latter case may result in extra scheduling cycles (even if we're going to fail them fast), but will reduce the overall scheduling time for the whole group - as we're waiting them to be all present in the queue first and then start the full scheduling cycle for each).

> Here we're adopting the latter approach, `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.

Reviewer 3: Are we still removing the PreFilter?

Author: I remove the mechanism to reject previous `PodGroup`, just left that ``PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.``

Reviewer 3: the total number of pods where? i.e. where does this counter get increased?

Author: > the total number of pods where? i.e. where does this counter get increased?
I get the total number of Pods belong to the same `PodGroup` through `PodLister.List(selector)` because they all have the same label `pod-group.scheduling.sigs.k8s.io/name` @alculquicondor

Reviewer 3: That would bypass all the caches, which may led to race conditions. Also it moves pods to the "unschedulable" queue if the number is lower than min. Then there is no way to bring them back unless some condition in the cluster changes that makes the pod "more schedulable". I know there is the flush every one minute, which I want to get rid of. But you probably don't want the plugin to have to wait 1min anyways.

Reviewer 4: > That would bypass all the caches

The PodLister should come from snapshot, so should be racing free. It's just that the `PodLister.List(filter)` does simple filtering instead of index, so perf can be improved.

> Then there is no way to bring them back unless some condition in the cluster changes that makes the pod "more schedulable".

I'm also concerned about this. However, think about a regular coscheduling implementation - i.e., which schedules every PodGroup Pod unconditionally and rely on Preempt plugin to wait to reach minAvailable, in that case, it seems this problem also exists?

Reviewer 3: That lister doesn't include unscheduled pods, which these are.

> However, think about a regular coscheduling implementation - i.e., which schedules every PodGroup Pod unconditionally and rely on Preempt plugin to wait to reach minAvailable, in that case, it seems this problem also exists?

That's not the case, because when a pod gets deleted, we bring back pods from the unschedulable queue back to the active queue.

Reviewer 4: > That lister doesn't include unscheduled pods, which these are.

That's true.

> That's not the case, because when a pod gets deleted, we bring back pods from the unschedulable queue back to the active queue.

But in both implementations, no (waiting) Pods get deleted - they just get moved from activeQ to unschedulableQ/packoffQ, or the other way around.

Reviewer 3: You said "Preempt plugin". I thought you were just referring to preemption. Are you referring to Reserve?
In that case, pods move to the plugin's cache. They are neither in active or unschedulableQ.

Reviewer 4: oops... my bad, typo, I meant "Permit" plugins.

Reviewer 3: No worries. So my comment still applies. We are not moving pods to unschedulable queue when we tell pods to "wait".

Reviewer 4: True. My point was that based on current extension points, this seems to be a common limitation among all coscheduling plugins.
